velox/exec/tests/RowContainerTest.cpp

```diff
@@ -32,44 +32,159 @@ using namespace facebook::velox::test;
+
+      std::vector<vector_size_t> rowNumbers;
+      rowNumbers.resize(size - offset);
+      for (int i = 0; i < size - offset; i++) {
+        rowNumbers[i] = i + offset;
       }
-    }
+      folly::Range<const vector_size_t*> rowNumbersRange =
+          folly::Range(rowNumbers.data(), size - offset);
+      container.extractColumn(
+          rows.data(), rowNumbersRange, column, offset, result);
+      testEqualVectors(result, expected, offset, offset);
+    };
+
+    // Test using extractColumn API.
+    testBasic();
+    testBasicWithOffset();
+    // Test using extractColumn (with rowNumbers) API.
+    testRowNumbers();
   }
 
-  void testExtractColumnForAllRows(
+  void testExtractColumnOutOfOrder(
       RowContainer& container,
       const std::vector<char*>& rows,
       int column,
-      const VectorPtr& expected) {
+      const VectorPtr& expected,
+      vector_size_t offset) {
     auto size = rows.size();
 
-    auto result = BaseVector::create(expected->type(), size, pool_.get());
-    container.extractColumn(rows.data(), size, column, result);
+    // In this test the extractColumn API is operated on an out-of-order
+    // input (either the base input data or the rowNumbers). The new ordering
+    // has null in even positions and retains odd positions from the original
+    // input.
+
+    // This is a helper method for result verification for the out of order
+    // results.
+    auto verifyResults = [&](const VectorPtr& result, int offset = 0) {
+      EXPECT_EQ(size, result->size());
+      for (vector_size_t i = offset; i < size; ++i) {
+        if (i % 2 == 0) {
+          EXPECT_TRUE(result->isNullAt(i)) << "at " << i;
+        } else {
+          EXPECT_TRUE(expected->equalValueAt(result.get(), i, i))
+              << "at " << i << ": expected " << expected->toString(i)
+              << ", got " << result->toString();
+        }
+      }
+    };
+
+    auto testBasic = [&]() {
+      // Construct an input source with null in even
+      // positions and that retains the original rows in
+      // odd positions. Copy its rows
+      std::vector<char*> input(size, nullptr);
+      for (auto i = 1; i < size; i += 2) {
+        input[i] = rows[i];
+      }
+
+      // Test the extractColumn API that didn't use the offset parameter
+      // and copies to the start of the result vector.
+      auto result = BaseVector::create(expected->type(), size, pool_.get());
+      container.extractColumn(input.data(), size, column, result);
+      verifyResults(result);
+
+      // Test extractColumn from offset.
+      container.extractColumn(
+          input.data() + offset, size - offset, column, offset, result);
+      verifyResults(result, offset);
+    };
+
+    auto testRowNumbers = [&]() {
+      // Setup input rows vector so that the first row is null.
+      std::vector<char*> input(rows);
+      input[0] = nullptr;
+
+      // The rowNumbersBuffer has values like 0, 1, 0, 3, 0, 5, 0, 7, etc
```

Comment:
The choice of the row numbers is a bit strange as it seems to test only writing nulls output of order and repeatedly. It doesn't seem to be testing writing non-null values out of order and/or repeatedly.

Reply:
I constructed that so that we retained only odd rows.

Don't want to disrupt this PR now if possible. I'll send out another PR once this lands that basically copies the value in the first position to all rows (using the rowNumbers API only). wdyt ?